### DISCUSSION TITLE
Allow async functions for key_func and cost

### DISCUSSION THREAD

**Discussion by boehm-e** [laurentS/slowapi#128] [Ideas]
I think it could be usefull to be able to call async function for key_func and cost.
It would allow us to get the Request data, and use that as the cost.
My usecase is the following:
- An anonymous user calls my webservice to process some text
- I want to allow him to process only 1000 chars / day.
- I would like to be able to use this library as follows:
```python

async def get_hit_cost(request: Request) -> int:
    payload = await request.json()
    return len(payload.text)

@app.post("/processText", status_code=200)
@limiter.limit("1000/day", cost=get_hit_cost)
async def getRecommandationsDemo(data: ProcessTextRequestModel, request: Request):
    return processedText;
```

**Comment by thentgesMindee** [comment]
actually, I think something even better than just support "async function", would be to have this package work as a dependency instead of a decorator, since it would allow us to get all advantages of fastapi's dependency injection system.
The issue with decorators here, is that since they are executed before any dependency, we cannot use anything computed in our dependency.

At my company, with one of our applications, we have a dynamic key_func based on something retrieved from one of our dependencies, and we encountered your issue. We found a very ugly workaround that works, but it would be waaay better to have it work as a dependancy directly, without having to do our weird workaround.

Let me know if I understood your use-case correctly, and if seeing parts of our workaround code could help you.

**Comment by boehm-e** [comment]
Yes, making this package work as a dependency would be great.
Yes @thentgesMindee I would like to see your workaround :)

**Reply by thentgesMindee** [reply]
Hey sorry about that, I was sure I answered and lost track of your message.. Finally found it back and actually gave my workaround there a few days ago: https://github.com/laurentS/slowapi/discussions/179#discussioncomment-8109168, hope it's not too late (almost one year..))
Let me know if it helped
